frontend/modules/teamspaces/teamspaces.redux.ts

```diff
@@ -0,0 +1,137 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { createActions, createReducer } from 'reduxsauce';
+import { keyBy } from 'lodash';
+
+export const { Types: TeamspacesTypes, Creators: TeamspacesActions } = createActions({
+	setTeamspaces: ['teamspaces'],
+	setModelUploadStatus: ['teamspace', 'project', 'model', 'status'],
+	// Projects
+	createProject: ['teamspace', 'projectData'],
+	updateProject: ['teamspace', 'projectName', 'projectData'],
+	removeProject: ['teamspace', 'projectName'],
+	createProjectSuccess: ['teamspace', 'projectData'],
+	updateProjectSuccess: ['teamspace', 'projectName', 'projectData'],
+	removeProjectSuccess: ['teamspace', 'projectName'],
+	// Models
+	createModel: ['teamspace', 'modelData'],
+	updateModel: ['teamspace', 'modelName', 'modelData'],
+	removeModel: ['teamspace', 'modelData'],
+	createModelSuccess: ['teamspace', 'modelData'],
+	updateModelSuccess: ['teamspace', 'modelName', 'modelData'],
+	removeModelSuccess: ['teamspace', 'modelData']
+}, { prefix: 'TEAMSPACES_' });
+
+export const INITIAL_STATE = {
+	teamspaces: []
+};
+
+const setTeamspaces = (state = INITIAL_STATE, action) => {
+	const teamspaces = keyBy(action.teamspaces, 'account');
+	return { ...state, teamspaces };
+};
+
+// Projects
+const updateProjectSuccess = (state = INITIAL_STATE, action) => {
+	const teamspaces = { ...state.teamspaces };
+	const projects = [...state.teamspaces[action.teamspace].projects].map((project) => {
+		if (project.name === action.projectName) {
+			return { ...project, ...action.projectData };
+		}
+		return project;
+	});
+	teamspaces[action.teamspace].projects = projects;
+
+	return { ...state, teamspaces };
+};
+
+const createProjectSuccess = (state = INITIAL_STATE, action) => {
+	const teamspaces = { ...state.teamspaces };
+	teamspaces[action.teamspace].projects.push(action.projectData);
+	return { ...state, teamspaces };
+};
+
+const removeProjectSuccess = (state = INITIAL_STATE, action) => {
+	const teamspaces = { ...state.teamspaces };
+	const projects = [...state.teamspaces[action.teamspace].projects]
+		.filter(({ name }) => name !== action.projectName);
+	teamspaces[action.teamspace].projects = projects;
+
+	return { ...state, teamspaces };
+};
+
+// Models
+const updateModelSuccess = (state = INITIAL_STATE, action) => {
+	const teamspaces = { ...state.teamspaces };
+	const projects = [...state.teamspaces[action.teamspace].projects];
+	const findedProject = projects.find((project) => project.name === action.modelData.projectName);
+	const projectIndex = projects.indexOf(findedProject);
+	const findedModel = findedProject.models.find((model) => model.name === action.modelData.name);
```

Comment:
```suggestion
	const findedModel = findedProject.models.find((model) => model.name === action.modelData.name);
```
```suggestion
	const modelIndex = foundProject.models.findIndex((model) => model.name === action.modelData.name);
```
You dont need `foundModel`